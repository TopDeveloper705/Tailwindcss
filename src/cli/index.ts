import arg from 'arg';
import { resolve } from 'path';
import { Processor } from '../lib';
import { readFileSync, writeFileSync } from 'fs';
import { HTMLParser } from '../utils/parser';
import { StyleSheet } from '../utils/style';
import {
  getVersion,
  FilePattern,
  walk,
  isFile,
  generateTemplate,
  Console,
} from './utils';

const doc = `
Generate css from text files that containing tailwindcss classes.
By default, it will use interpretation mode to generate a single css file.

Usage:
  windicss [filenames]
  windicss [filenames] -c -m
  windicss [filenames] -c -s -m
  windicss [filenames] [-c | -i] [-b | -s] [-m] [-p <prefix:string>] [-o <path:string>] [--args arguments]

Options:
  -h, --help            Print this help message and exit.
  -v, --version         Print windicss current version and exit.

  -i, --interpret       Interpretation mode, generate class name corresponding to tailwindcss. This is the default behavior.
  -c, --compile         Compilation mode, combine the class name in each row into a single class.
  -t, --preflight       Add preflights, default is false.

  -b, --combine         Combine all css into one single file. This is the default behavior.
  -s, --separate        Generate a separate css file for each input file.

  -m, --minify          Generate minimized css file.
  -p, --prefix PREFIX   Set the css class name prefix, only valid in compilation mode. The default prefix is 'windi-'.
  -o, --output PATH     Set output css file path.
  -f, --config PATH     Set config file path.

  --init PATH           Start a new project on the path.
`;

const args = arg({
  // Types
  '--help': Boolean,
  '--version': Boolean,
  '--compile': Boolean,
  '--interpret': Boolean,
  '--preflight': Boolean,
  '--combine': Boolean,
  '--separate': Boolean,
  '--minify': Boolean,
  '--init': String,
  '--prefix': String,
  '--output': String,
  '--config': String,

  // Aliases
  '-h': '--help',
  '-v': '--version',
  '-i': '--interpret',
  '-c': '--compile',
  '-t': '--preflight',
  '-b': '--combine',
  '-s': '--separate',

  '-m': '--minify',
  '-p': '--prefix',
  '-o': '--output',
  '-f': '--config',
});

if (args['--help'] || (args._.length === 0 && Object.keys(args).length === 1)) {
  Console.log(doc);
  process.exit();
}

if (args['--version']) {
  Console.log(getVersion());
  process.exit();
}

if (args['--init']) {
  const template = generateTemplate(args['--init'], args['--output']);
  args._.push(template.html);
  args['--preflight'] = true;
  args['--output'] = template.css;
}

const localFiles = walk('.', true).filter((i) => i.type === 'file').map(i=> ({ type: i.type, path: i.path.replace(/\\/g, '/') }) );

const matchFiles: string[] = [];
for (const pt of args._) {
  if (isFile(pt) && pt.search(/\.windi\./) === -1) {
    matchFiles.push(pt);
  } else if (pt.search(/\*/) !== -1) {
    // match files like **/*.html **/src/*.html *.html ...
    localFiles.forEach((i) => {
      const pattern = new FilePattern(pt);
      if (pattern.match(i.path) && i.path.search(/\.windi\./) === -1)
        matchFiles.push(i.path);
    });
  } else {
    Console.error(`File ${pt} does not exist!`);
  }
}

let ignoredClasses: string[] = [];
const preflights: StyleSheet[] = [];
const styleSheets: StyleSheet[] = [];
const processor = new Processor(args['--config'] ? require(resolve(args['--config'])) : undefined);

if (args['--compile']) {
  // compilation mode
  const prefix = args['--prefix'] ?? 'windi-';
  matchFiles.forEach((file) => {
    let indexStart = 0;
    const outputStyle: StyleSheet[] = [];
    const outputHTML: string[] = [];
    const html = readFileSync(file).toString();
    const parser = new HTMLParser(html);

    // Match tailwind ClassName then replace with new ClassName
    parser.parseClasses().forEach((p) => {
      outputHTML.push(html.substring(indexStart, p.start));
      const utility = processor.compile(p.result, prefix, true); // Set third argument to false to hide comments;
      outputStyle.push(utility.styleSheet);
      ignoredClasses = [...ignoredClasses, ...utility.ignored];
      outputHTML.push([utility.className, ...utility.ignored].join(' '));
      indexStart = p.end;
    });
    outputHTML.push(html.substring(indexStart));
    styleSheets.push(
      outputStyle.reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
    );

    const outputFile = file.replace(/(?=\.\w+$)/, '.windi');
    writeFileSync(outputFile, outputHTML.join(''));
    Console.log(`${file} -> ${outputFile}`);

    if (args['--preflight'])
      preflights.push(processor.preflight(parser.html));
  });
} else {
  // interpretation mode
  matchFiles.forEach((file) => {
    const parser = new HTMLParser(readFileSync(file).toString());
    const utility = processor.interpret(
      parser
        .parseClasses()
        .map((i) => i.result)
        .join(' ')
    );
    styleSheets.push(utility.styleSheet);
    ignoredClasses = [...ignoredClasses, ...utility.ignored];

    if (args['--preflight'])
      preflights.push(processor.preflight(parser.html));
  });
}

if (args['--separate']) {
  styleSheets.forEach((style, index) => {
    const filePath = matchFiles[index].replace(/\.\w+$/, '.windi.css');
    if (args['--preflight']) style = preflights[index].extend(style);
    writeFileSync(filePath, style.build(args['--minify']));
    Console.log(`${matchFiles[index]} -> ${filePath}`);
  });
} else {
  let outputStyle = styleSheets
    .reduce(
      (previousValue: StyleSheet, currentValue: StyleSheet) =>
        previousValue.extend(currentValue),
      new StyleSheet()
    )
    .combine()
    .sort();
  if (args['--preflight'])
    outputStyle = preflights
      .reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
      .combine()
      .sort()
      .extend(outputStyle);
  const filePath = args['--output'] ?? 'windi.output.css';
  writeFileSync(filePath, outputStyle.build(args['--minify']));
  Console.log('matched files:', matchFiles);
  Console.log('output file:', filePath);
}

Console.log('ignored classes:', ignoredClasses);
