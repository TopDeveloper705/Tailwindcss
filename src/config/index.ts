export { baseConfig } from './base';
export { colors } from './colors';
export const twExclude = [
  /-hex-/,                          // disable hex color
  /-\$/,                            // disable variable
  /(rem|em|px|rpx|vh|vw|ch|ex)$/,   // disable size
  /-\d*[13579]$/,                   // disable odd number
  /-([0-9]{1,}[.][0-9]*)$/,         // disable float
  /^!/,                             // disable important utility
];
