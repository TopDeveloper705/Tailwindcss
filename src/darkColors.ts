export { darkColors as default } from './config';
