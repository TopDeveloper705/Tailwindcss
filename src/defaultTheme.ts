import { Processor } from "./lib";

export default new Processor().allTheme;
