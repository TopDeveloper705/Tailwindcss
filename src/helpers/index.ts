export * from './template';
export * from './transform';
