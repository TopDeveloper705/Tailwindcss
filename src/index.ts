export * from './processor';
export * from './utils';
export * from './config';
export * from './plugin';

module.exports = {
    'colors': require('./config/colors')
}

