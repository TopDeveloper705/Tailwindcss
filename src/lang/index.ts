export { default as Lexer } from './lexer';
export { default as Parser } from './parser';
export { default as Interpreter } from './interpreter';

export { evaluate as eval } from './eval';
