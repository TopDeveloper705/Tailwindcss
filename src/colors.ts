import { default as colors } from './config/colors';

module.exports = colors;