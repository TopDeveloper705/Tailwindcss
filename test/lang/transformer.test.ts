import { Lexer, Parser, Transformer } from '../../src/lang';

describe('Transformer', () => {
  it('simple syntax', () => {
    const code = `
      @import 'typo.css', 'style.css';

      @load 'module1', 'module2', 'module3';
      @load { export1 } from "module-name";
      @load { export1 , export2 } from "module-name";
      @load { export1 , export2 as alias2 , export3 as alias3 } from "module-name";
      @load * from "module";
      @load * as name from "module";
      @load defaultExport from "module";
      @load defaultExport, { export1, export2 } from "module-name";
      @load defaultExport, * as name from 'module-name';

      @func func(a, b) {
        @var f = 3;
        @func nest(c, d) {
          @return c * d;
        };
        @return a + b / f - nest(a, b);
      };
      @func add(a, b) => a + b;
      @func (a, b) => a + b;
      @func (a, b) {
        @return a + b;
      };
      @func (x) {
        @return x + 2;
      };

      @try {
        @var x = 1;
      } @except TypeError {
        @log 'TypeError: Failed to set x';
      } @except RangeError {
        @log 'RangeError: Failed to set x';
      } @except Exception as error {
        @log \`Other exceptions \${error.message}\`;
      } @else {
        @log 'No exception occured';
      } @finally {
        @log 'We always do this';
      };

      @try {
        fail();
      } @except {
        @log 'Exception occured';
      };

      @try {
        result = 10 / x;
      } @except TypeError as error1 {
        @log \`Type Error \${error1.message}\`;
      };

      @try {
        result = 10 / x;
      } @except Exception as error {
        @log 'Exceptions';
      };

      @var add3 = a => a + 3;

      @var add = (a, b) => a + b;

      @var c = ((a, b) => a + b)(1,2) * 4;

      @var d = ((a, b) => {
        @return a + b;
      })(1, 2) / 3;

      @var number = 2;
      @var a = number;
      @var b = 10 * a + 10 * number / 4;
      @var c = a - - b;
      @var d = 'green';
      @var e = [ 10, 'green' , a - b ];
      @var f = ( 10, 'green' , a - b );
      @var g = {
        1: 2,
        'a': 1,
        'b': 2,
        'c': a - b,
      };

      @if a > 0 and b > 0 {
        @log "The numbers are greater than 0";
      };

      @if a > 0 and b > 0 and c > 0 {
        @log "The numbers are greater than 0";
      } @else {
        @log "Atleast one number is not greater than 0";
      };

      @if a > 0 {
        @log "a";
      } @elif b > 0 {
        @log "b";
      } @elif c > 0 {
        @log "c";
      } @else {
        @log "unknown";
      };

      @var i = 3;
      @while i > 0 {
        @log i;
        i = i - 1;
      } @else {
        @log i;
      };

      @with add(1,2) as c {
        c = c + 3;
        @log c;
      };

      @break;
      @continue;
      @yield 123;
      @raise Error('msg');

      @assert not 1 == 2 and 5 > 3;

      @assert !1 == 2 and 5 > 3;

      @var e = a.c(1, 2)('5').hello()(123)['test'].def;

      @assert 4 in [1, 2, 4];

      @var h = func2(func(1, 2) + (4 / 3));

      @var h = True;
      @var i = False * 3;
      @var j = None;

      c = - 4 ** 4;

      a + 4;

      @log d;

      @log b;

      @warn a + 12;

      @apply bg-red-500 text-white sm:bg-gray-200;
      @attr[bg] red-500 opacity-50 sm:red-200;
      @attr[sm:text] red-500 opacity-30;

      @js {
        import { eval, rgba, get, set } from 'windi/lang';

        const a = get('width');
        set('width', eval('4px'));
        const width = eval('3px');
        export function add(a, b) {
          return a + b;
        }
      }

      .test {
        @var testNest = 123;
        color: red;
        .abc .def {
          @apply font-bold text-lg;
          color: \${d};
        }
      }

      .def {
        @var testNest = "green";
        color: red;
        &:hover > nest {
          background: \${testNest};
        }
      }
    `;
    const lexer = new Lexer(code);
    const parser = new Parser(lexer);
    const transformer = new Transformer(parser);
    expect(transformer.transform()).toMatchSnapshot('js');
  });
});
