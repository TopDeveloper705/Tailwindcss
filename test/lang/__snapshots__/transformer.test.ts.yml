Tools / simple syntax / js / 0: |-
  __import__("typo.css");
  __import__("style.css");
  import "module1";
  import "module2";
  import "module3";
  import { export1 } from "module-name";
  import { export1, export2 } from "module-name";
  import { export1, export2 as alias2, export3 as alias3 } from "module-name";
  import * from "module";
  import * as name from "module";
  import defaultExport from "module;
  import defaultExport, { export1, export2 } from "module-name";
  import defaultExport, * as name from "module-name";
  function func(a, b) {
  let f = 3;
  function nest(c, d) {
  return __mul__(c, d);
  };
  return __minus__(__add__(a, __div__(b, f)), nest(a, b));
  };
  const add = (a, b) => __add__(a, b);
  ((a, b) => __add__(a, b));
  (function (a, b) {
  return __add__(a, b);
  });
  (function (x) {
  return __add__(x, 2);
  });
  async function f() {
  let promise = new Promise((function (resolve, reject) {
  setTimeout((() => resolve("done!")), 1000);
  }));
  let result = await promise;
  alert(result);
  };
  try {
    let x = 1;
  console.log("No exception occured");
  }catch(e) {
    if (e instanceof TypeError) {
    console.log("TypeError: Failed to set x");
  } else if (e instanceof RangeError) {
    console.log("RangeError: Failed to set x");
  } else if (e) {
    let error = e;
  console.log(`Other exceptions ${error.message}`);
  }} finally {
    console.log("We always do this");
  };
  try {
    fail();
  }catch(e) {
      console.log("Exception occured");};
  try {
    result = __div__(10, x);
  }catch(e) {
    if (e instanceof TypeError) {
    let error1 = e;
  console.log(`Type Error ${error1.message}`);
  }};
  try {
    result = __div__(10, x);
  }catch(e) {
    if (e) {
    let error = e;
  console.log("Exceptions");
  }};
  let add3 = ((a) => __add__(a, 3));
  let add = ((a, b) => __add__(a, b));
  let c = __mul__(((a, b) => __add__(a, b))(1, 2), 4);
  let d = __div__((function (a, b) {
  return __add__(a, b);
  })(1, 2), 3);
  let number = 2;
  let a = number;
  let b = __add__(__mul__(10, a), __div__(__mul__(10, number), 4));
  let c = __minus__(a, __negative__(b));
  let d = "green";
  let e = [10, "green", __minus__(a, b)];
  let f = [10, "green", __minus__(a, b)];
  let g = {  
    1: 2,
    "a": 1,
    "b": 2,
    "c": __minus__(a, b)
  };
  if (a > 0 && b > 0) {
    console.log("The numbers are greater than 0");
  };
  if (a > 0 && b > 0 && c > 0) {
    console.log("The numbers are greater than 0");
  } else {
    console.log("Atleast one number is not greater than 0");
  };
  if (a > 0) {
    console.log("a");
  } else if (b > 0) {
    console.log("b");
  } else if (c > 0) {
    console.log("c");
  } else {
    console.log("unknown");
  };
  let i = 3;
  while (i > 0) {
    console.log(i);
  i = __minus__(i, 1);
  };
  if (!(i > 0)) {
    console.log(i);
  };
  () => {
    let c = add(1, 2);
    c = __add__(c, 3);
  console.log(c);
  };
  break;
  continue;
  yield 123;
  throw Error("msg");
  console.assert(!(1 === 2) && 5 > 3);
  console.assert(!(1) === 2 && 5 > 3);
  let e = a.c(1, 2)("5").hello()(123)["test"].def;
  console.assert(__in__(4, [1, 2, 4]));
  let h = func2(__add__(func(1, 2), __div__(4, 3)));
  let h = true;
  let i = __mul__(false, 3);
  let j = undefined;
  let date = __add__(new Date("123"), __div__(new Date("456"), 4));
  c = __negative__(__exp__(4, 4));
  __add__(a, 4);
  console.log(d);
  console.log(b);
  console.warn(__add__(a, 12));
  delete a;
  __interpret__("bg-red-500 text-white sm:bg-gray-200");
  __attributify__("bg", "red-500 opacity-50 sm:red-200");
  __attributify__("sm:text", "red-500 opacity-30");

          import { eval, rgba, get, set } from 'windi/lang';

          const a = get('width');
          set('width', eval('4px'));
          const width = eval('3px');
          export function add(a, b) {
            return a + b;
          }
        ;
  (() => {
  const style = new Style(".test");
  let testNest = 123;
  style.add(new Property("color", "red"));
  style.add((() => {
  const style = new Style(".abc .def");
  __interpret__("font-bold text-lg");
  style.add(new Property("color", `${d}`));
  return style;
  })());
  return style;
  })();
  (() => {
  const style = new Style(".def");
  let testNest = "green";
  style.add(new Property("color", "red"));
  style.add((() => {
  const style = new Style("&:hover > nest");
  style.add(new Property("background", `${testNest}`));
  return style;
  })());
  return style;
  })();
